service/worker/scanner/executions/mutable_state_validator.go

```diff
@@ -213,3 +235,35 @@ func (v *mutableStateIDValidator) validateID(
 ) bool {
 	return common.FirstEventID <= eventID && eventID <= lastEventID
 }
+
+func (v *mutableStateIDValidator) validateRetention(
+	executionInfo *persistencespb.WorkflowExecutionInfo,
+	executionState enums.WorkflowExecutionState,
+) (*MutableStateValidationResult, error) {
+	if executionState != enums.WORKFLOW_EXECUTION_STATE_COMPLETED {
+		return nil, nil
+	}
+	// We don't use the close time here because some old workflows do not have the close time.
+	// It makes sense the workflow is finished and use the last update time.
+	finalUpdateTime := executionInfo.GetLastUpdateTime()
+	if finalUpdateTime == nil {
+		return nil, nil
+	}
+
+	ttl := time.Now().UTC().Sub(timestamp.TimeValue(finalUpdateTime))
+	ns, err := v.registry.GetNamespaceByID(namespace.ID(executionInfo.GetNamespaceId()))
+	if err != nil {
+		return nil, err
+	}
+	retention := ns.Retention()
+	if ttl > 0 && ttl > retention {
```

Comment:
Do we want to add some buffer here in case deletion time is not created/executed due to some other tasks pending (like archival task)?

Reply:
Yes. Add a buffer config